src/Components/Web.Extensions/src/InputFile/RemoteBrowserFileStream.cs

```diff
@@ -0,0 +1,153 @@
+            _maxChunkSize = maxChunkSize;
+
+            var pipe = new Pipe(new PipeOptions(pauseWriterThreshold: maxBufferSize, resumeWriterThreshold: maxBufferSize));
+            _pipeReader = pipe.Reader;
+            _fillBufferCts = new CancellationTokenSource();
+
+            _ = FillBuffer(pipe.Writer, _fillBufferCts.Token);
+        }
+
+        private async Task FillBuffer(PipeWriter writer, CancellationToken cancellationToken)
+        {
+            long offset = 0;
+
+            while (offset < File.Size)
+            {
+                var pipeBuffer = writer.GetMemory(_maxChunkSize);
+                var chunkSize = (int)Math.Min(_maxChunkSize, File.Size - offset);
+
+                try
+                {
+                    var base64 = await _jsRuntime.InvokeAsync<string>(
+                        InputFileInterop.ReadFileData,
+                        cancellationToken,
+                        _inputFileElement,
+                        File.Id,
+                        offset,
+                        chunkSize);
+
+                    if (!Convert.TryFromBase64String(base64, pipeBuffer.Span, out var bytesWritten))
+                    {
+                        throw new FormatException("A chunk with an invalid format was received.");
+                    }
+
+                    if (bytesWritten != chunkSize)
+                    {
+                        throw new InvalidOperationException(
+                            $"A chunk with size {bytesWritten} bytes was received, but {chunkSize} bytes were expected.");
+                    }
+
+                    writer.Advance(chunkSize);
+                    offset += chunkSize;
+                }
+                catch (Exception e)
+                {
+                    await writer.CompleteAsync(e);
+                    throw;
+                }
+
+                var result = await writer.FlushAsync();
+
+                if (result.IsCompleted)
+                {
+                    break;
+                }
+            }
+
+            await writer.CompleteAsync();
+        }
+
+        protected override async ValueTask<int> CopyFileDataIntoBuffer(long sourceOffset, Memory<byte> destination, CancellationToken cancellationToken)
+        {
+            if (_isReadingCompleted)
+            {
+                return 0;
+            }
+
+            int totalBytesCopied = 0;
+
+            while (destination.Length > 0)
+            {
+                var result = await _pipeReader.ReadAsync(cancellationToken);
+
+                if (result.IsCanceled)
+                {
+                    _isReadingCompleted = true;
+
+                    var exception = new OperationCanceledException(cancellationToken);
+                    await _pipeReader.CompleteAsync(exception);
+
+                    throw exception;
+                }
+
+                var bytesToCopy = (int)Math.Min(result.Buffer.Length, destination.Length);
```

Comment:
Can we bail if this evaluates to 0? The way this is currently set up this could infinite loop when `result.IsCompleted == false`